src/js/stores/MesosStore.js

```diff
@@ -0,0 +1,325 @@
+  addDataToMap(requestId, requestMap, requestData);
+  requestCallback();
+}
+
+function updateRequest(requestId, nextRequestData) {
+  var requestData = getDataFromMap(requestId, requestMap,
+    REQUEST_DATA_TTL);
+  if (requestData != null) {
+    addDataToMap(requestId, requestMap,
+      Object.assign(requestData, nextRequestData));
+  }
+}
+
+function resetRequest(requestId) {
+  invalidateMapData(requestId, requestMap);
+}
+
+function resolveFileRequest(fileRequest, queueIndex) {
+  MesosStore.emit(MesosEvents.REQUEST_TASK_FILES_COMPLETE, fileRequest);
+  taskFileRequestQueue.splice(queueIndex, 1);
+}
+
+function rejectFileRequest(fileRequest, queueIndex) {
+  MesosStore.emit(MesosEvents.REQUEST_TASK_FILES_ERROR, fileRequest);
+  taskFileRequestQueue.splice(queueIndex, 1);
+}
+
+function resolveTaskFileRequests() {
+  var info = InfoStore.info;
+
+  if (taskFileRequestQueue.length === 0) {
+    return;
+  }
+
+  if (!info) {
+    performRequest(INFO_ID,
+      () =>  InfoActions.requestInfo(),
+      () => taskFileRequestQueue.forEach(rejectFileRequest));
+    return;
+  }
+
+  if (!Util.isString(info.frameworkId) ||
+      !Util.isObject(info.marathon_config) ||
+      !Util.isString(info.marathon_config.mesos_leader_ui_url)) {
+    taskFileRequestQueue.forEach(rejectFileRequest);
+    return;
+  }
+
+  resetRequest(INFO_ID);
+
+  if (!version) {
+    MesosActions.requestVersionInformation(
+      info.marathon_config.mesos_leader_ui_url.replace(/\/$/, ""));
+    return;
+  }
+
+  if (!MesosStore.getState(MASTER_ID)) {
+
+    performRequest(MASTER_ID,
+      () => MesosActions.requestState(MASTER_ID,
+        info.marathon_config.mesos_leader_ui_url.replace(/\/?$/, "/master"),
+        version),
+      () => taskFileRequestQueue.forEach(rejectFileRequest));
+    return;
+  }
+
+  taskFileRequestQueue.forEach((fileRequest, queueIndex) => {
+    var taskId = fileRequest.taskId;
+    var agentId = fileRequest.agentId;
+
+    if (!MesosStore.getState(agentId)) {
+      let masterState = MesosStore.getState(MASTER_ID);
+      let nodeURL = getNodeURLFromState(agentId, masterState);
+      if (nodeURL == null) {
+        invalidateMapData(MASTER_ID, stateMap);
+        resolveTaskFileRequests();
+        return;
+      }
+
+      performRequest(agentId,
+        () => MesosActions.requestState(agentId, nodeURL, version),
+        () => rejectFileRequest(fileRequest, queueIndex));
+      return;
+    }
+
+    resetRequest(MASTER_ID);
+
+    if (!MesosStore.getTaskFiles(taskId)) {
+      let masterState = MesosStore.getState(MASTER_ID);
+      let agentState = MesosStore.getState(agentId);
+      let nodeURL = getNodeURLFromState(agentId, masterState);
+      let executorDirectory =
+        getExecutorDirectoryFromState(info.frameworkId, taskId, agentState);
+
+      if (nodeURL == null || executorDirectory == null) {
+        invalidateMapData(agentId, stateMap);
+        resolveTaskFileRequests();
+        return;
```

Comment:
In my configuration the following happens on user click:

1. `MesosActions.requestTaskFiles(task.slaveId, task.id);`
2. `MesosEvents.REQUEST_TASK_FILES` gets dispatched
3. the taskId gets pushed to the `taskFileRequestQueue` by the `MesosStore` and `resolveTaskFileRequests()` is called
4.  `InfoStore.info` is there, skip first check (line 186)
5. the 2nd check is also skipped as the info is there (lines 193 and on)
6. `resetRequest(INFO_ID);` is called
7. `version` is undefined, we enter the block at line 202, and trigger `MesosActions.requestVersionInformation`
8. In my setup, the `/version` endpoint fails, triggering a reject on the JSONP promise
9. `REQUEST_VERSION_INFORMATION_ERROR` is triggered
10. we enter `resolveTaksFileRequest` again
11. this time `version` is "0.0.0"
12. ...
13. we end up in this block and `rejectFileRequest` is never called.

When the above happens, `REQUEST_TASK_FILES_ERROR` never gets triggered.